### PR TITLE
mwan3: fix interface-bound traffic when interface is offline

### PR DESCRIPTION
**Pull request by marcin1j**
Maintainer: @feckert 
Compile tested: ar71xx, NETGEAR 3700v2
Run tested: ar71xx, NETGEAR 3700v2

Description:

This commit fixed what 6d99b602 was supposed to fix without affecting interface-bound traffic.

Before 6d99b602 interface-bound traffic was working normally as long as at least one interface was  online. However when the last interface went offline, it was impossible to ping and such state was unrecoverable.

Commit 6d99b602 fixed unrecoverable offline state problem (it was possible to ping -I iface) but messed inteface-bound traffic. Traffic with interface source address was not working if the interface was in
"offline" state, even if another interface was online.
The problem was caused by an inconsistent "offline" interface state: iptables-related rules were kept while routing table and policy were deleted.

The idea behind this commit is to:
 1. Keep all the rules for each interface (iptables, routing table, policy) regardless of its state. This ensures consistency,
 2. Make interface state hotplug events affect only iptables'  `mwan3_policy_*` rules. Interface-related iptables, routing table and policy is removed only when mwan3 is manually stopped.

To make such changes possible, it's necessary to change the way `mwan3_policy_*` rule generator keeps track of interface state hotplug events.
Until now, it checked for the existence of custom interface-related routing table (table id 1, 2, 3, ...). Clearly we can no longer rely on that so each interface state is stored explicitly in file.

### CODE REVIEW

File: net/mwan3/Makefile [hunk 1]
```diff
@@ -9,7 +9,7 @@ include $(TOPDIR)/rules.mk
 
 PKG_NAME:=mwan3
 PKG_VERSION:=2.6.3
-PKG_RELEASE:=2
```
Comment (feckert) [inline net/mwan3/Makefile]: Could you please update this to PKG_RELEASE 1?
Because this is a source change an not a Makefile change.

File: net/mwan3/files/lib/mwan3/mwan3.sh [hunk 2]
```diff
@@ -784,16 +797,14 @@ mwan3_report_iface_status()
 
 	if [ -z "$id" -o -z "$device" ]; then
 		result="unknown"
-	elif [ -n "$($IP rule | awk '$1 == "'$(($id+1000)):'"')"i -a -n "$($IP rule | awk '$1 == "'$(($id+2000)):'"')" -a -n "$($IPT -S mwan3_iface_in_$1 2> /dev/null)" -a -n "$($IPT -S mwan3_iface_out_$1 2> /dev/null)" -a -n "$($IP route list table $id default dev $device 2> /dev/null)" ]; then
-		result="online"
+	elif [ -n "$($IP rule | awk '$1 == "'$(($id+1000)):'"')" -a -n "$($IP rule | awk '$1 == "'$(($id+2000)):'"')" -a -n "$($IPT -S mwan3_iface_in_$1 2> /dev/null)" -a -n "$($IPT -S mwan3_iface_out_$1 2> /dev/null)" -a -n "$($IP route list table $id default dev $device 2> /dev/null)" ]; then
```
Comment (feckert) [inline net/mwan3/files/lib/mwan3/mwan3.sh]: Could you please move this fix to a separate commit so i could cherry-pick this to lede-17.01 because i have seen it must be fixed there as well!!
This was my fault!! Thanks for fixing!

**Comment by feckert** [comment]
I will test the changes and if everything works as expected we could merge this.
This is a really smart solution! Nice
If we merge this we should cherry-pick the fix to lede-17.01

File: net/mwan3/Makefile [hunk 3]
```diff
@@ -9,7 +9,7 @@ include $(TOPDIR)/rules.mk
 
 PKG_NAME:=mwan3
 PKG_VERSION:=2.6.3
```
Comment (feckert) [inline net/mwan3/Makefile]: Could you please update this to PKG_VERSION 2.6.4?
Because this is a source change an not a Makefile change.

File: net/mwan3/files/lib/mwan3/mwan3.sh [hunk 4]
```diff
@@ -8,15 +8,15 @@ IPT6="ip6tables -t mangle -w"
 LOG="logger -t mwan3[$$] -p"
 CONNTRACK_FILE="/proc/net/nf_conntrack"
 
-MWAN3_STATUS_DIR="/var/run/mwan3track"
+MWAN3_STATUS_DIR="/var/run/mwan3"
 
+[ -d $MWAN3_STATUS_DIR ] || mkdir $MWAN3_STATUS_DIR $MWAN3_STATUS_DIR/state
```
Comment (feckert) [inline net/mwan3/files/lib/mwan3/mwan3.sh]: Could you use 
[ -d $MWAN3_STATUS_DIR ] || mkdir **-p** $MWAN3_STATUS_DIR/state

File: net/mwan3/files/usr/sbin/mwan3 [hunk 5]
```diff
@@ -160,7 +159,7 @@ stop()
 	done
 
 	mwan3_lock_clean
-	rm -rf "${MWAN3_STATUS_DIR}/mmx_mask"
+	rm -rf $MWAN3_STATUS_DIR
```
Comment (feckert) [inline net/mwan3/files/usr/sbin/mwan3]: We should also delete /var/run/mwan3track if mwan3 is stopped
rm -rf $MWAN3TRACK_STATUS_DIR

This variable could be defined in **/lib/mwan3/mwan3.sh** after
https://github.com/marcin1j/openwrt-packages/blob/80fb9ca76fbbc5baf0a50dd4861f0e0614f0c027/net/mwan3/files/lib/mwan3/mwan3.sh#L11

**Comment by marcin1j** [comment]
Hi

Sorry, I've been offline for a few days again. Now PR is updated and your comments disappeared.

One thing that cannot be done is separating this commit for cherry pick. By doing that you would introduce an error in `lede-17.01` branch and `/usb/sbin/mwan3` would always show interface `online`.
So backport needs this line manually merged and this I can prepare similar PR for `lede-17.01` branch.

**Comment by feckert** [comment]
@marcin1j thanks for your contribution :-)

> So backport needs this line manually merged and this I can prepare similar PR for lede-17.01 branch.

This would be nice if we could apply this fix also to lede-17.01. This should replace the hotfix
https://github.com/openwrt/packages/commit/282e90014e936ed1db527608e6ac4904a895b1ab#diff-34590d11545c84b961d1992ce3aefdeb

@hnyman we can merge this thanks

**Comment by marcin1j** [comment]
@feckert 
Adjusting this commit to LEDE 17.01 needed a couple of conflicts resolved manually (more than one would expect). Anyway the backport  is ready. It seems to work but the catch is that I could test it only on a single device.
Do you have enough devices to test this thoroughly?

Should I open new PR for the backport (if its too early, you can pick commit marcin1j@2c90c5d from my repository)?

**Comment by feckert** [comment]
@marcin1j we should also delete this line
https://github.com/marcin1j/openwrt-packages/blob/2c90c5d68a850b0be4bcbcc5c62d4680d9d1d461/net/mwan3/files/usr/sbin/mwan3#L49

which was set by this hotfix
https://github.com/marcin1j/openwrt-packages/commit/282e90014e936ed1db527608e6ac4904a895b1ab

**Comment by feckert** [comment]
@marcin1j i have tested this on my lantiq/xrx200 and it works as expected.

If last interface has lost connectivity only the unreachable (https://github.com/marcin1j/openwrt-packages/blob/df1f3a41c8935386d37c02303843f1cc232d620b/net/mwan3/files/lib/mwan3/mwan3.sh#L530) target is set for this policies.
And if the interface gets back online my rules are setup again.
So it recovers from offline state if last interface could ping the tracking ips again.

**Comment by marcin1j** [comment]
@feckert 

You're right, I've left this line by mistake:
```
mwan3_delete_iface_iptables $1
```

Having fixed that I've tested everything once again (marcin1j/openwrt-packages@4739584). I guess now it's time to open PR against `lede-17.01` branch.
